### PR TITLE
Consistently set label on axis with units

### PR DESCRIPTION
**Pull request by dstansby**
## PR Summary
Fixes https://github.com/matplotlib/matplotlib/issues/23416. The issue was an incorrect check on whether units needed to be updated.

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -1610,8 +1610,17 @@ def _update_axisinfo(self):
         self.set_default_intervals()
 
     def have_units(self):
+        """
+        Return `True` if units or a converter have been set.
+        """
         return self.converter is not None or self.units is not None
 
+    def have_units_and_converter(self):
+        """
```
Comment (jklymak) [inline lib/matplotlib/axis.py:1619]: Should we make this private so we can more easily change the future api?  Units are likely to get an overhaul with the new NASA Roses funding so I would be conservative about adding new units API just now.

**Review by timhoffm — DISMISSED** [review]
I'm convinced this is the wrong way. See https://github.com/matplotlib/matplotlib/issues/23416#issuecomment-1182476487. Please discuss there. I'm blocking this until we've reached a conclusion there.

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (dstansby) [reply lib/matplotlib/axis.py:1619]: 👍 done

File: lib/matplotlib/tests/test_units.py [hunk 2]
```diff
@@ -283,3 +283,23 @@ def test_plot_kernel():
...
+    ax_ref = fig_ref.subplots()
+    axs = [ax_test, ax_ref]
+
+    for ax in axs:
+        ax.yaxis.set_units('km')
+        ax.set_xlim(10, 20)
+        ax.set_ylim(10, 20)
+
+    ax_test.scatter([1, 2, 3], data, edgecolors='none')
+    ax_ref.plot([1, 2, 3], data, marker='o', linewidth=0)
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_units.py:305]: We can directly test the label value (and on both axis):

```suggestion
    fig, ax = plt.subplots()
    ax.xaxis.set_units('m')
    ax.yaxis.set_units('km')
    ax.scatter(np.arange(3) * units.m, np.arange(3) * units.km)
    ax.get_xlabel() == 'm'
    ax.get_ylabel() == 'km'
```

Maybe add suplots for different plot-functions

File: lib/matplotlib/axes/_base.py [hunk 3]
```diff
@@ -2549,7 +2549,9 @@ def _process_unit_info(self, datasets=None, kwargs=None, *, convert=True):
             except KeyError:
                 raise ValueError(f"Invalid axis name: {axis_name!r}") from None
             # Update from data if axis is already set but no unit is set yet.
-            if axis is not None and data is not None and not axis.have_units():
+            if (axis is not None and
+                    data is not None and not
+                    axis._have_units_and_converter()):
```
Comment (timhoffm) [inline lib/matplotlib/axes/_base.py:2554]: Is this only fixing symptoms but not the root cause?

At this position `plot()` results in `units` and `converter` being set, but `scatter()` has only set `units` and `converter` is still None. As little as I know about units - this discrepancy feels wrong.

Comment (dstansby) [reply lib/matplotlib/axes/_base.py:2554]: (it's been a while but...) I think that inside this `if` clause is designed to ensure that both units and a converter have been set, so it makes sense alter the logic to do this if either one is not set.

File: lib/matplotlib/tests/test_units.py (hunk 2 above)
Comment (dstansby) [reply lib/matplotlib/tests/test_units.py:305]: 👍 test updated

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/axes/_base.py [hunk 4]
```diff
@@ -2580,7 +2580,9 @@ def _process_unit_info(self, datasets=None, kwargs=None, *, convert=True):
...
```
Comment (QuLogic) [inline lib/matplotlib/axes/_base.py:2585]:
```suggestion
                    data is not None and
                    not axis._have_units_and_converter()):
```

**Review by QuLogic — APPROVED** [review]

**Pull request merged**